File: April2021/PartitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    // TC : O(n)
    // SC  : O(1)
    public ListNode partition(ListNode head, int x) {
        ListNode smallerHead = new ListNode(-1);
        ListNode greaterHead = new ListNode(-1);
        ListNode sIt = smallerHead;
        ListNode gIt = greaterHead;

        ListNode it = head;
        while(it!=null){
            if(it.val <x){
                // for smaller values
                sIt.next = it;
                it = it.next;
                sIt =sIt.next;
            } else{
                // for equal or greater comparisons
                gIt.next = it;
                it = it.next;
                gIt = gIt.next;
            }
        }

        sIt.next = greaterHead.next;
        gIt.next = null;

        return smallerHead.next;

    }
}

File: August2023 /Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void setPointers(ListNode*&  s,ListNode*& fs,ListNode*& head){
        if(!s){
            s=head;
            fs=s;    
        } 
        else{
            s->next=head;
            s=s->next;
        }
    }
    ListNode* partition(ListNode* head, int x) {
        ListNode* s=NULL,*fs=NULL,*b=NULL,*fb=NULL;
        while(head){
            if(head->val<x) setPointers(s,fs,head);
            else setPointers(b,fb,head);
            head=head->next;
        }
        if(s) s->next=fb;
        if(b) b->next=NULL;
        return fs?fs:fb;
    }
};